async_hooks: use parent promise as triggerId in init hook

JiaLiPassion opened the pull request:
async_hooks init callback will be triggered when promise newly created,
in previous version, the parent promise which pass from chrome V8 PromiseHook
is ignored, so we can't tell the promise is a pure new promise or a
chained promise.

In this commit, we use the parent promise's id as triggerId to
trigger the init callback.

Fixes: https://github.com/nodejs/node/issues/13302

Thanks for @AndreasMadsen and @addaleax 's great help! 

Please review.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->

JiaLiPassion commented on src/async-wrap.cc:
```diff
@@ -296,7 +296,17 @@ static void PromiseHook(PromiseHookType type, Local<Promise> promise,
+      // current promise's triggerId
+      double trigger_id = Unwrap<PromiseWrap>(parent.As<Promise>())->get_id();
+      env->set_init_trigger_id(trigger_id);
+      wrap = new PromiseWrap(env, promise, silent);
```
@AndreasMadsen , Thank very much for the review, I will try to understand this case and your solution.

JiaLiPassion commented:
@AndreasMadsen , about the sample code you provided,

```
Consider the following code:

const p0 = new Promise((resolve) => resolve(1));
const p1 = p0.then(function () {
  // async_hooks.currentId() === asyncId(p1)
  p2 = p1.then(function () { // trigger_id = asyncId(p1)
    
  });
  return 2;
});
```

I make a test with this PR compiled node.

```
const async_hooks = require('async_hooks');

function init(id, provider, parentId, parentHandle) {
    process._rawDebug('init: ', id, 'triggerId: ', parentId, 'currentId',  async_hooks.currentId());
}

function before(id) { 
    process._rawDebug('before', id, async_hooks.currentId());
 }
function after(id) { 
    process._rawDebug('after', id, async_hooks.currentId());
 }
function destroy(id) {
    process._rawDebug('destroy', id);
}

async_hooks.createHook({init, before, after, destroy}).enable();
debugger;
const p = new Promise ( resolve => {
    process._rawDebug('execute p');
    resolve(1);} );
const p1 = p.then(val => { 
    process._rawDebug('p1 then', val); 
    const p2 = p1.then((val1) => {
      process._rawDebug('p2 then', val1); 
    });
    return 2; });
```

the output is 

```
init:  2 triggerId:  1 currentId 1
execute p
init:  3 triggerId:  2 currentId 1
before 3 1
p1 then 1
init:  4 triggerId:  3 currentId 1
after 3 1
before 4 1
p2 then 2
after 4 1
```

p is promise without parent, so `triggerId === currentId`
p1 is promise ,  and p1's parent is p, so the triggerId is 2
p2 is promise ,  and p2's parent is p1, so the triggerId is 3,

I think this a correct result.

And I tried another case

```
setTimeout(() => {
    const p = new Promise(resolve => {
        process._rawDebug('execute p');
        resolve(1);
    });
    const p1 = p.then(val => {
        process._rawDebug('p1 then', val);
        const p2 = p1.then((val1) => {
            process._rawDebug('p2 then', val1);
        });
        return 2;
    });

}, 100);
```

the output is 

```
init:  2 type: Timeout triggerId:  1 currentId 1
init:  3 type: TIMERWRAP triggerId:  1 currentId 1
before 3 3
before 2 2
init:  4 type: PROMISE triggerId:  2 currentId 2
execute p
init:  5 type: PROMISE triggerId:  4 currentId 2
after 2 2
after 3 3
before 5 0
p1 then 1
init:  6 type: PROMISE triggerId:  5 currentId 0
after 5 0
before 6 0
p2 then 2
after 6 0
destroy 2
```

I think the result is also ok, but I don't quite understand the why `currentId` will become `0`.

Please review. Thank you very much!

AndreasMadsen commented:
Interesting, I think something is wrong with `async_hooks.currentId()`.

```diff
init:  2 triggerId:  1 currentId 1
execute p
init:  3 triggerId:  2 currentId 1
before 3 1
p1 then 1
init:  4 triggerId:  3 currentId 1 // wrong, currentId should be 3
after 3 1
before 4 1
p2 then 2
after 4 1
```

/cc @addaleax @trevnorris

JiaLiPassion commented:
@AndreasMadsen, I am not sure I understand the lifecycle of currentId, so I do not know what is the correct result,just like the another sample involve setTimeout will have a currentId 0,I do not understand why.

AndreasMadsen commented:
@JiaLiPassion The `async_hooks.currentId()` should just be the `id` from the latest `before` emit that hasn't been closed by the `after` emit.

In some cases, there is no `before` emit (and it is also not the root) then `currentId` will be `0`, but that is not the case in either examples.

JiaLiPassion commented:
@AndreasMadsen , Thank you for the explanation.

>  The async_hooks.currentId() should just be the id from the latest before emit that hasn't been closed by the after emit.

Yeah, this is making sense.

> In some cases, there is no before emit (and it is also not the root) then currentId will be 0, but that is not the case in both examples.

you mean there is only `init` emit, is that right?

And yes, in my setTimeout example, there is `before` emit, so I don't know why there is a `0` currentId.  ( I have tried the nightly build version of nodejs, the `currentId` result is the same).

AndreasMadsen commented:
> you mean there is only init emit, is that right?

For example in the TCP server you will get:

```
init TCP socket
before TCP server onconnection
....
after TCP server onconnection
```

This is because the TCP socket is created in C++ land and then send to the callback. Since this happens before any javascript the `currentId` is zero.

addaleax commented on src/async-wrap.cc:
```diff
@@ -296,6 +296,14 @@ static void PromiseHook(PromiseHookType type, Local<Promise> promise,
+      // parent promise exists, current promise
+      // is a chained promise, so we set parent promise's id as
+      // current promise's triggerId
+      double trigger_id = Unwrap<PromiseWrap>(parent.As<Promise>())->get_id();
```
You need to also check whether `Unwrap<PromiseWrap>()` returned `nullptr` (that could happen if `async_hooks` were enabled after `parent` was created, but before `promise` was created).

If it is, you’ll need to basically do the same thing we’re doing just below; something along the lines of:

```c++
auto parent_wrap = new PromiseWrap(env, parent, true);
parent_wrap->MakeWeak(parent_wrap);
```

JiaLiPassion replied:
@addaleax , thank you for the review.

So the code should look like 

```
      auto parent_wrap = Unwrap<PromiseWrap>(parent.As<Promise>());
      if (parent_wrap != nullptr) {
         double trigger_id = parent_wrap->get_id();
         env->set_init_trigger_id(trigger_id);
      }
 
```

addaleax replied:
@JiaLiPassion I think you would still want to create a *new* `PromiseWrap` for the `parent` Promise. If you do that, you don’t need to check `if (parent_wrap != nullptr)`.

(Basically, look at the existing code here and how it adds a `parent_wrap` when there isn’t any. If it’s unclear, feel free to ping me and I’ll explain in more detail. :))

JiaLiPassion replied:
@addaleax , thank you so much! sorry for so many basic questions...

So the code should like 

```
// create a new PromiseWrap for parent promise
auto parent_wrap = new PromiseWrap(env, parent, true);
parent_wrap->MakeWeak(parent_wrap);
// get id from parentWrap
double trigger_id = parent_wrap->get_id();
env->set_init_trigger_id(trigger_id);
```

addaleax replied:
The new `PromiseWrap` should only be created if `Unwrap<>` gives you a `nullptr` (otherwise there already is one, so we don’t need that). But yes, that’s the general idea.

JiaLiPassion replied:
@addaleax , thank you, so it will look like

```
auto parent_wrap = Unwrap<PromiseWrap>(parent.As<Promise>());
if (parent_wrap == nullptr) {
  // create a new PromiseWrap for parent promise with silent parameter
  parent_wrap = new PromiseWrap(env, parent, true);  
  parent_wrap->MakeWeak(parent_wrap);
}
// get id from parentWrap
double trigger_id = parent_wrap->get_id();
env->set_init_trigger_id(trigger_id);
```

AndreasMadsen commented:
> The code change looks good to me, aside from one edge case. I still am not 100 % sure about this being the right thing, but if Andreas says so, I trust him.

Sorry to burst your bubble, but the `triggerId` is something I really struggle with. I tried to explain some of my confusion in https://github.com/nodejs/node/pull/11883#pullrequestreview-36665607. Unfortunately, I still haven't gotten any wiser regarding when to set `triggerId` and when to let it default to `currentId`.

Also since this doesn't solve the problem that @JiaLiPassion originally asked, regarding detecting if a parentPromise exists, I would like to explore some alternatives.

tl;dr; I think this makes sense, but I'm generally uncertain about `triggerId`.

> I would still really like to see a test that shows that getting the original execution scope remains possible, because that is something that actual use cases will rely on.

Yeah, we need to fix the `currentId` bug before that can happen. I think the test case that should follow with that should be sufficient.

addaleax commented:
@AndreasMadsen Regarding `triggerId`, I can definitely see that there are at least two valid perspectives:

- If it’s about tracking asynchronous *causality*, this PR **might** be the right one. It would make a Promise look like it descended from where the origin of the Promise chain was created. I am not sure that’s super helpful, though; a Promise can be resolved with another Promise, breaking the causality chain, and afaik we can’t track that well currently.
- If it’s about tracking asynchronous *context*, this PR is not the right one; basically, if you want to have the code after an `await` in an async fn look like it runs in a context descended from the one before `await`, then the current behaviour is what you want.

Writing this comment, it still seems to me like we should maybe just expose the parent Promises async id as an extra argument to the `init` callback.

AndreasMadsen commented:
@addaleax It is the first one, but there is a thin line between causality and parent resource, which is what causes my confusion.

From the documentation:

> `triggerId` is the `asyncId` of the resource that caused (or "triggered") the
new resource to initialize and that caused `init` to call. This is different
from `async_hooks.currentId()` that only shows *when* a resource was created,
while `triggerId` shows *why* a resource was created.

JiaLiPassion commented:
@AndreasMadsen , @addaleax , 

> Writing this comment, it still seems to me like we should maybe just expose the parent Promises async id as an extra argument to the init callback.

So if expose the parent Promise async id as an extra argument, it can resolve my original issue, that if the argument exists, the the promise is a chained one, is that correct?


> If it’s about tracking asynchronous context, this PR is not the right one; basically, if you want to have the code after an await in an async fn look like it runs in a context descended from the one before await, then the current behaviour is what you want.

do you mean the case like this?

```
async function test() {
    await new Promise(resolve => resolve(1));
}

process._rawDebug('before', async_hooks.currentId());
test();
process._rawDebug('after', async_hooks.currentId());

```

addaleax commented:
@AndreasMadsen Okay, makes sense.

> So if expose the parent Promise async id as an extra argument, it can resolve my original issue, that if the argument exists, the the promise is a chained one, is that correct?

@JiaLiPassion Yes, I’d say so.

> do you mean the case like this?

More like this:

```js
async function test() {
  process._rawDebug('before', async_hooks.currentId());
  await new Promise(resolve => resolve(1));
  process._rawDebug('after', async_hooks.currentId());
}
```

AndreasMadsen commented:
> Writing this comment, it still seems to me like we should maybe just expose the parent Promises async id as an extra argument to the init callback.

I think extra arguments to `init` just for the purpose of promises is overdoing it. That additional API will no purpose for any other resource in `async_hooks` but promises. If those informations are just necessary for promises then I think a documented `PromiseWrap` resource is the way to go.

addaleax commented:
> If those informations are just necessary for promises then I think a documented `PromiseWrap` resource is the way to go.

That works for me; but it means that we’d have to switch the `resource` argument that’s passed to `init` from the promise itself to a distinct object, right?

AndreasMadsen commented:
> That works for me; but it means that we’d have to switch the resource argument that’s passed to init from the promise itself to a distinct object, right?

I think that would be the best. It is also what we discussed in the original Promise PR and what @Fishrock123 advocated for.

JiaLiPassion commented:
@AndreasMadsen , sorry I just lost...

> That works for me; but it means that we’d have to switch the resource argument that’s passed to init from the promise itself to a distinct object, right?
> I think that would be the best. It is also what we discussed in the original Promise PR and what @Fishrock123 advocated for.

Do you mean use `AsyncResource` in `async_hook.js` to pass the `parent promise` information?

AndreasMadsen commented:
> Do you mean use AsyncResource in async_hook.js to pass the parent promise information?

Sort of, but we definitely shouldn't go through JS for that. For the purpose of a `resource` object, `AsyncResource` is just an object. The `resource` object should look something like:

```js
resource = {
  promise // reference to promise,
  parentPromise // reference to parentPromise or undefined
}
```

JiaLiPassion commented:
@AndreasMadsen , can I just change the how to create a new  `PromiseWrap` to something like

```
    wrap = new PromiseWrap(env, {
      promise,
      parentPromise
    }, silent);
 
```

AndreasMadsen commented:
@JiaLiPassion I don't think that is enough. This is as far as my V8-C++ knowledge goes, I'm not sure how to tie in the GC of `promise` with `promise_resource` (this is important for emitting `destroy`) and I don't know the most performant way of doing this.

trevnorris commented:
This conversation is progressing fast, so hopefully this isn't out of context.

@AndreasMadsen 
> Sorry to burst your bubble, but the triggerId is something I really struggle with.

Sorry. Let me try to clear up when to propagate a different `triggerId`. First, we need to separate the value returned by `async_hooks.triggerId()` and `triggerId` passed to `init()`. They cover two different ideas.

First let's talk about `currentId()` and `triggerId()`:
```js
const s = net.createServer(onconnection).listen(0, function() {
  // The .listen() callback is called from either a nextTick() or cares lookup() and
  // the currentId() will reflect that, but since the callback is being called from the
  // TCP instance it's triggerId() should reflect this.
  assert(async_hooks.currentId() > 0 &&
         async_hooks.triggerId() !== s._handle.getAsyncId());
  assert(async_hooks.triggerId() === s._handle.getAsyncId());
});
```
So in the case of a `Promise`, this is what I'd expect:
```js
let cid = null;
new Promise(res => {
  cid = async_hooks.currentId();
}).then(val => {
  assert(async_hooks.triggerId() === cid);
  cid = async_hooks.currentId();
}).catch(e => {
  assert(async_hooks.triggerId() === cid);
});
```
<br/>

Now about when JS resources are meant to emulate native resource for reasons like pooling JS resources onto a single native resource. An example of this in core is timers. Multiple JS `Timer` instances are attached to a single `TimerWrap` instance:
```js
const sym = Object.getOwnPropertySymbols(setTimeout(() => {}))
  .filter(i => i.toString() === 'Symbol(asyncId)')[0];
const s1 = setTimeout(() => {}, 100);
const s2 = setTimeout(() => {}, 100);
assert(s1._idleNext._timer === s2._idlePrev._timer);
assert(s1[sym] !== s1._idleNext._timer.getAsyncId());
assert(s2[sym] !== s2._idlePrev._timer.getAsyncId());
assert(s1[sym] !== s2[sym]);
```
<br/>

This could also be used by something like a database module, where multiple queries should be treated as their own async operations, but the actual operation is combined into a single query internally. This is the reason for `async_hooks.runInAsyncIdScope()`:
```js
// A pooled resource on top of a native resource
class MyResource inherits async_hooks.AsyncResource {
  constructor() {
    super('MyResource');
  }
  query(val, cb) {
    // query() will pool the actual request then call each callback with
    // the specific data it requested.
    nativeResource.query(val, (er, data) => {
      async_hooks.runInAsyncIdScope(this.asyncId(), () => cb(er, data));
    });
  }
}

const mr = new MyResource();
mr.query(val, (er, data) => {
  // Now when init() fires for writeFile() it will have both the currentId and
  // triggerId of the JS instance.
  fs.writeFile(path, data.toString());
});
```

<br/>

I'll write up more examples to more thoroughly explain this and hopefully make it a more pragmatic process of determining when to propagate `triggerId`.

JiaLiPassion commented:
@trevnorris , thank you  very much for the comment, I think it will take some time to understand what you described.

for this case, 

1. do you think pass `parent promise` asyncId as `triggerId` for `promise` in `init` callback is correct?
2. could you help to provide a way in `async_hooks` callback to identify whether a promise has `parent promise` or not?

JiaLiPassion commented:
@AndreasMadsen , I will continue to learn how to pass the `resource` object, is there any sample about how to do that? I searched `TCPWrap` and `FSEventWrap` , but have not figured out how to do that.

AndreasMadsen commented:
@trevnorris Thanks for the explanation, it mostly confirms my current understanding. The DB queue example is a little odd, as far as I know, nowhere in node-core do we manipulate triggerId this way. Would the following approach not be more inline with how Timers works?

```js
class RequestResource inherits async_hooks.AsyncResource {
  constructor(val, triggerId) {
    super('RequestResource', triggerId);
    this.val = val;
  }
}

// A pooled resource on top of a native resource
class PoolResource inherits async_hooks.AsyncResource {
  constructor() {
    super('PoolResource');
  }
  query(val, cb) {
    // query() will pool the actual request then call each callback with
    // the specific data it requested.
    const requestResource = new RequestResource(val, this.asyncId());
    nativeResource.query(val, function (er, data) {
      requestResource.emitBefore();
      cb(er, data);
      requestResource.emitAfter();
      requestResource.emitDestroy();
    });
  }
}
```

> do you think pass parent promise asyncId as triggerId for promise in init callback is correct?

based on the promise example @trevnorris provided I would say yes.

> could you help to provide a way in async_hooks callback to identify whether a promise has parent promise or not?

This is a good question.

@JiaLiPassion could you provide a detailed explanation for why this information is necessary for zone.js? I'm not convinced it is.

JiaLiPassion commented:
@AndreasMadsen , in `zone.js`, it basically do the same work as `async_hooks` or Domain, the basic usage will be 

```
Zone.current.fork(
  {
     name: 'testZone',
     properties: {context: { key : 'value'}}, // some context value will be shared through all async operations in the same zone
     onScheduleTask: function(delegate, targetZone, currentZone, Task)  {
       // schedule a async task, this is very similar to init callback in async_hooks
       return delegate.scheduleTask(targetZone, task);
     }, 
     onInvokeTask: function(delegate, targetZone, currentZone, applyThis, applyArgs) {
        // here is the position  just like before callback in async_hooks
        const r delegate.invokeTask(targetZone, task, applyThis, applyArgs);
        // here is the position  just like after callback in async_hooks
        return r;
     }
  }
).run(() => {
   const p = new Promise ( resolve => {
    process._rawDebug('execute p');
    resolve(1);} );
const p1 = p.then(val => { 
    process._rawDebug('p1 then', val); 
    const p2 = p1.then((val1) => {
      process._rawDebug('p2 then', val1); 
    });
    return 2; });
 
});

```

And basically, `onScheduleTask` is just like `init` of `async_hooks`, and `onInvokeTask` is just like `before` + `after` of `async_hooks`.

But there is a difference is , in `zone.js`, schedule a task means there will be an async operation.
maybe a macroTask(setTimeout) or a microTask (Promise.then), maybe it will not be invoked (such as cancelled or error), but there should be a async operation happen, so inside `zone.js`,  there will be `counter` to record there is a new async Task will occur and in `scheduled` state.

So this behavior is different with `init`, especially in `Promise` case.

for example, 

```
const p = new Promise ( resolve => {
    process._rawDebug('execute p');
    resolve(1);} );
```

**This sample will not trigger any async operation, it just create a promise with `resolved` state.
But in `async_hooks`, there will be `init` callback, in `zone.js`, `onScheduleTask` will not be triggered **

```
const p = new Promise ( resolve => {
    process._rawDebug('execute p');
    resolve(1);} );

const p1 = p.then(r => ...);
```

When `Promise.then` is called, there will be an async microTask being scheduled, so  both `init` and `onScheduleTask` will be triggered.

So I need to know a promise `init` callback is an async operation or not.

Please help me to check the idea is correct or not , thanks!

AndreasMadsen commented:
@JiaLiPassion I see, that makes sense. You definitely need to know if `parentPromise` was set then to filter `init`, the resource object will solve that.

For the causal context, the `triggerId` solution you have implemented here will solve that.

JiaLiPassion commented:
@AndreasMadsen , Thank you very much!

So , beside the changes of this PR.

1. so the `async_hooks.currentId()` is still not correct as you pointed?
2. we still need to have a resource object to pass to init, this is a better solution , right? but just like you said, we need to make the resource object can be GC by v8.

AndreasMadsen commented:
>  we still need to have a resource object to pass to init, this is a better solution , right? but just like you said, we need to make the resource object can be GC by v8.

It is another piece of the solution. About the GC: when the `promise` object is GCed the `destroy` hook is emitted. If we pass in the `PromiseWrap` object the `destroy` event will be when the `PromiseWrap` is GCed. We don't want this, because if the user doesn't use the `PromiseWrap` resource object that can happen too early.

JiaLiPassion commented:
@AndreasMadsen , got it, I will try to learn about it.

With this PR and the current `async_hooks.currentId()` behavior, I can check parent promise with 
`triggerId == async_hooks.currentId()`, but if the `async_hooks.currentId()` changed(current version it seems the currentId will not change when Promise.then occurs), then we have to implement promise resource object to resolve this issue, is that right?

trevnorris commented:
@AndreasMadsen 
> The DB queue example is a little odd

That's because it is. Wrote those examples in a hurry and messed that up. I'm writing more concrete rules for `currentId()` and `triggerId()` and will post it here.

AndreasMadsen commented:
`triggerId == async_hooks.currentId()` will never work, no matter how few bugs there are in `async_hooks.currentId()`. This is due to the edge-case I described in https://github.com/nodejs/node/pull/13367#pullrequestreview-41485890.

AndreasMadsen commented:
> Aside: The discussion here is moving pretty fast. Going to grab some sleep and will read over everything in the morning and respond.

I can summarize it for you.

1. The `triggerId` should be as you described. This PR fixes that. Except in your example `async_hooks.currentId();` in `new Promise` is `1` (root). I also think this is fine.

```js
const p0 = new Promise(res => {
  async_hooks.currentId(); // this is not equal to asyncId(p0), but asyncId(root)
});
const p1 = p0.then(val => {
  assert(async_hooks.triggerId() === asyncId(p0));
});
const p1.catch(e => {
  assert(async_hooks.triggerId() === asyncid(p1));
});
```

2. The `async_hooks.currentId()` appear to be very bugged. See https://github.com/nodejs/node/pull/13367#issuecomment-305494868

3. We should change the resource object for promises such it is:

```js
resource = {
  promise // reference to promise,
  parentPromise // reference to parentPromise or undefined
}
```

and not the promise itself. This will allow the user to detect if `parentPromise` was set. This is required for implementing zone.js

AndreasMadsen commented:
@JiaLiPassion I think we should merge this as is and solve the other problems separately. But we need some more testing.

Could you write a test for:

```diff
+      if (parent_wrap == nullptr) {
+        // create a new PromiseWrap for parent promise with silent parameter
+        parent_wrap = new PromiseWrap(env, parent_promise, true);
+        parent_wrap->MakeWeak(parent_wrap);
+      }
```

To test it you need to:

1. create a `p0 = new Promise` that resolves immediately.
2. create AsyncHook listeners.
3. call `p1 = p0.then()` that resolves immediately.
4. ensure the correct events fires with the correct values (´p0.destroy, p1.init, p1.before, p1.after, p1.destroy´).

Also please add a test for:

1. create a `p0 = new Promise` that resolves immediately.
2. call `p0.then()` that resolves immediately.
3. create AsyncHook listeners.
4. ensure the correct events fires with the correct values (´p0.destroy, p1.before, p1.after, p1.destroy´).

JiaLiPassion commented:
@AndreasMadsen ,sure , I will add those cases.

JiaLiPassion commented:
@AndreasMadsen , I have added the test cases, the behaviors are different with you described.

1. Case1 , initHooks after new Promise

create a p0 = new Promise that resolves immediately.
create AsyncHook listeners.
call p1 = p0.then() that resolves immediately.
events fires with the values (p1.init, p1.before, p1.after).

2. Case2, initHooks after Promise.then

create a p0 = new Promise that resolves immediately.
call p0.then() that resolves immediately.
create AsyncHook listeners.
events fires with the values (empty).

- In the previous `test-promise.js` the `destroy` was not assert because the GC timing is not controlled. 

- And in case2, the initHooks was created after Promise.then, so it should be empty.

Please review.

AndreasMadsen commented:
Thanks for writing the tests.

> In the previous test-promise.js the destroy was not assert because the GC timing is not controlled.

True.

> And in case2, the initHooks was created after Promise.then, so it should be empty.

False. Future events should not depend on previous events. I can't really see what the bug could be.

JiaLiPassion commented:
@AndreasMadsen , I just update the cases.

> And in case2, the initHooks was created after Promise.then, so it should be empty.

This is not true, but in `init_hooks.js`, because there is no `init` callback, so we can't get those activities through `type` name `PROMISE`.

I tried to find the activity from `Unknown` activities and check the status.

```
2. Case2, initHooks after Promise.then

create a p0 = new Promise that resolves immediately.
call p0.then() that resolves immediately.
create AsyncHook listeners.
events fires with the values (p1.before, p1.after).
```

Please review.

AndreasMadsen commented:
Well done with the debugging!

We need to make sure the `triggerId` is actually correct and not just a number. Since we can't get the `id` of the parent dynamically (because there is no init event), we will have to just set it to a static value. Because this method isn't robust to changes, you should split the promise tests into 3 files, one for each test.

If we just check that `triggerId` is a number the test will pass without this PR (except the first one).

addaleax commented:
@JiaLiPassion just fyi, you can add syntax highlighting to your code by starting blocks with ```` ```js ````, ```` ```c++ ````, etc. instead of ```` ``` ````. I’ve edited some of your posts above while catching up on the thread so that they are more readable, I hope that’s okay. :)

addaleax commented:
@AndreasMadsen @JiaLiPassion I lost track a bit. Can you outline what we have left to do here? Can I help? If it’s about turning the `resource` into an object with `promise` and `parentPromise` properties, I can do that if you like.

AndreasMadsen commented:
@addaleax I have summarized what needs to be done in https://github.com/nodejs/node/pull/13367#issuecomment-305734393

1. In this PR we are going to solve part 1, set the `triggerId`.

2. Part 2, `async_hooks.currentId()` is wrong, is not really related to this but a bug we discovered.

3. To get zone.js working we need the `resource` object (part 3). I'm not sure how to do that correctly, we would appreciate some help.

JiaLiPassion commented:
@AndreasMadsen , thank you for review, I will split the files and check the triggerId.

@addaleax , thank you, I will try to format my code better from now on.

> @JiaLiPassion just fyi, you can add syntax highlighting to your code by starting blocks with ```js, ```c++, etc. instead of ```. I’ve edited some of your posts above while catching up on the thread so that they are more readable, I hope that’s okay. :)

Sure, please.

> @AndreasMadsen @JiaLiPassion I lost track a bit. Can you outline what we have left to do here? Can I help? If it’s about turning the resource into an object with promise and parentPromise properties, I can do that if you like.

Yes, that will help!!!, I have no idea how to do it correctly, and I need it to implement `zone.js`

JiaLiPassion commented:
@AndreasMadsen , I just modified the test cases.

1. split the cases into 3 files

2. in case1, promise before `initHooks`, check the triggerId with a static value.

Please review.

hayes commented:
Reading through this, it looks like there were a couple of issues identified that won't be fixed by this PR

1. `async_hooks.currentId()` does not return the correct value while executing a PromiseReactionJob (maybe other cases?)  (see https://github.com/nodejs/node/pull/13367#issuecomment-305505712 and https://github.com/nodejs/promises/issues/9#issuecomment-305924526)
2. `async_hooks.triggerId()` is not set as expected to the id at the time of resolution. see https://github.com/nodejs/node/pull/13367#issuecomment-305647169

Should I create new issues to track these?

AndreasMadsen commented:
> async_hooks.currentId() does not return the correct value while executing a PromiseReactionJob (maybe other cases?) see #13367 (comment)

You can open an issue.

> async_hooks.triggerId() is not set as expected to the id at the time of resolution. see #13367 (comment) and nodejs/promises#9 (comment)

I'm not convinced there is an issue after this lands and the `async_hooks.currentId()` issue is fixed.

AndreasMadsen commented on test/async-hooks/test-promise.promise-before-init-hooks.js:
```diff
@@ -0,0 +1,44 @@
+  const a0 = as[0];
+  assert.strictEqual(a0.type, 'PROMISE', 'promise request');
+  assert.strictEqual(typeof a0.uid, 'number', 'uid is a number');
+  assert.strictEqual(a0.triggerId, 5, 'use parent uid as triggerId');
```
hmm, not sure what is going on. This should have `a0.triggerId = 2`.

JiaLiPassion replied:
I have modified it, when I run the case with `node test/async_hooks/test-promise.promise-before-init-hooks.js`, the `uid` of the parent promise will be `5`.

But when run `make -j4 test`, it will be `2`.

I will try to figure out why.

AndreasMadsen commented:
> in case1, promise before initHooks, check the triggerId with a static value.

Sorry for the misunderstanding, if you can check it with a dynamic value then that is much better. The static value is just a callback. Also, it doesn't appear that `test/async-hooks/test-promise.js` (case 1) is changed at all.

We should also do like https://github.com/nodejs/node/pull/13243, and not use unhelpful assert messages.

JiaLiPassion commented:
@AndreasMadsen , Thanks, I have 

1. updated the test cases to check the `triggerId` with dynamic parent promise uid.
 
2. and modify the assert messages.

Please review.

hayes commented:
> I'm not convinced there is an issue after this lands and the async_hooks.currentId() issue is fixed.
This goes back to a question that has come up many times in the context of tracing/apm.  There are 2 competing ideas about the best way to propagate state for promises, which stem from different use cases.  This change helps improve the first use case (zone.js) where you want your data flow to closely model the users code (state is propagated from the point a callback is attached to where the callback is run).

The second use case is important when you want to track causality (eg. your response time is slow because of a specific database query).  APM agents usually build up a tree of nodes for a given transaction,  With the current solution the following code:
```javascript
request1().then(() => {
  return request2()
}).then(() => {
  return request3()
})
```
would result in a shallow tree like:
```js
{
  name: 'root',
  children: [
    {
      name: 'request1',
      start: 0,
      end: 1000
    },
    {
      name: 'request2',
      start: 1000,
      end: 2000
    },
    {
      name: 'request3',
      start: 2000,
      end: 3000
    }
  ]
}
```

what you want instead for apm is to understand the causality between the various promises, and build a tree that looks something like:
```js
{
  name: 'root',
  children: [{
    name: 'request1',
    start: 0,
    end: 1000,
    children: [{
      name: 'request2',
      start: 1000,
      end: 2000,
      children: [{
        name: 'request3',
        start: 2000,
        end: 3000
      }]
    }]
  }]
}
```
According to https://github.com/AndreasMadsen/node/blob/82149a8e0f0a13e5f34215e370b6750a300617e6/doc/api/async_hooks.md#triggerid

triggerId should be the id of the resource that triggered the current callback to execute, which in this example would be the previous request (when it finishes and resolves the promise).  I believe this would need to be implemented by hooking into the promiseResove hook here: https://github.com/JiaLiPassion/node/blob/61c6d225352eecbc1c863a3b2e88307af79d28cc/src/async-wrap.cc#L319

hayes commented:
quick point of clarification, setting the triggered as I am proposing here does not prevent any existing behavior, it simply adds the ability to link the execution of a callback back to the async resource that caused it to be scheduled.

addaleax commented:
The diff for fancy resource objects should look something like this:

<details><summary>look in the fold</summary>

```diff
diff --git a/src/async-wrap.cc b/src/async-wrap.cc
index e2717c034473..817613f4c6a9 100644
--- a/src/async-wrap.cc
+++ b/src/async-wrap.cc
@@ -36,6 +36,7 @@ using v8::Context;
 using v8::Float64Array;
 using v8::Function;
 using v8::FunctionCallbackInfo;
+using v8::FunctionTemplate;
 using v8::HandleScope;
 using v8::HeapProfiler;
 using v8::Integer;
@@ -44,6 +45,7 @@ using v8::Local;
 using v8::MaybeLocal;
 using v8::Number;
 using v8::Object;
+using v8::ObjectTemplate;
 using v8::Promise;
 using v8::PromiseHookType;
 using v8::RetainedObjectInfo;
@@ -293,7 +295,13 @@ static void PromiseHook(PromiseHookType type, Local<Promise> promise,
                         Local<Value> parent, void* arg) {
   Local<Context> context = promise->CreationContext();
   Environment* env = Environment::GetCurrent(context);
-  PromiseWrap* wrap = Unwrap<PromiseWrap>(promise);
+  Local<Value> resource_object_value = promise->GetInternalField(0);
+  Local<Object> resource_object;
+  PromiseWrap* wrap = nullptr;
+  if (resource_object_value->IsObject()) {
+    resource_object = resource_object_value.As<Object>();
+    wrap = Unwrap<PromiseWrap>(resource_object);
+  }
   if (type == PromiseHookType::kInit || wrap == nullptr) {
     bool silent = type != PromiseHookType::kInit;
     // set parent promise's async Id as this promise's triggerId
@@ -302,18 +310,33 @@ static void PromiseHook(PromiseHookType type, Local<Promise> promise,
       // is a chained promise, so we set parent promise's id as
       // current promise's triggerId
       Local<Promise> parent_promise = parent.As<Promise>();
-      auto parent_wrap = Unwrap<PromiseWrap>(parent_promise);
+      Local<Value> parent_resource_value = parent_promise->GetInternalField(0);
+      Local<Object> parent_resource;
+      PromiseWrap* parent_wrap = nullptr;
+      if (parent_resource_value->IsObject()) {
+        parent_resource = parent_resource_value.As<Object>();
+        parent_wrap = Unwrap<PromiseWrap>(parent_resource);
+      }
 
       if (parent_wrap == nullptr) {
+        parent_resource = env->promise_wrap_template()
+                             ->NewInstance(context).ToLocalChecked();
+        parent_resource->Set(context, env->promise_string(), parent).FromJust();
+        parent_promise->SetInternalField(0, parent_resource);
         // create a new PromiseWrap for parent promise with silent parameter
-        parent_wrap = new PromiseWrap(env, parent_promise, true);
+        parent_wrap = new PromiseWrap(env, parent_resource, true);
         parent_wrap->MakeWeak(parent_wrap);
       }
       // get id from parentWrap
       double trigger_id = parent_wrap->get_id();
       env->set_init_trigger_id(trigger_id);
     }
-    wrap = new PromiseWrap(env, promise, silent);
+    resource_object = env->promise_wrap_template()
+                         ->NewInstance(context).ToLocalChecked();
+    resource_object->Set(context, env->promise_string(), promise).FromJust();
+    resource_object->Set(context, env->parent_string(), parent).FromJust();
+    promise->SetInternalField(0, resource_object);
+    wrap = new PromiseWrap(env, resource_object, silent);
     wrap->MakeWeak(wrap);
   } else if (type == PromiseHookType::kResolve) {
     // TODO(matthewloring): need to expose this through the async hooks api.
@@ -353,6 +376,15 @@ static void SetupHooks(const FunctionCallbackInfo<Value>& args) {
   SET_HOOK_FN(destroy);
   env->AddPromiseHook(PromiseHook, nullptr);
 #undef SET_HOOK_FN
+
+  {
+    Local<FunctionTemplate> ctor =
+        FunctionTemplate::New(env->isolate());
+    ctor->SetClassName(FIXED_ONE_BYTE_STRING(env->isolate(), "PromiseWrap"));
+    Local<ObjectTemplate> promise_wrap_template = ctor->InstanceTemplate();
+    promise_wrap_template->SetInternalFieldCount(1);
+    env->set_promise_wrap_template(promise_wrap_template);
+  }
 }
 
 
diff --git a/src/env.h b/src/env.h
index c8c8232cc07f..dbf1999e2cab 100644
--- a/src/env.h
+++ b/src/env.h
@@ -189,12 +189,14 @@ namespace node {
   V(parse_error_string, "Parse Error")                                        \
   V(path_string, "path")                                                      \
   V(pbkdf2_error_string, "PBKDF2 Error")                                      \
+  V(parent_string, "parent")                                                  \
   V(pid_string, "pid")                                                        \
   V(pipe_string, "pipe")                                                      \
   V(port_string, "port")                                                      \
   V(preference_string, "preference")                                          \
   V(priority_string, "priority")                                              \
   V(produce_cached_data_string, "produceCachedData")                          \
+  V(promise_string, "promise")                                                \
   V(raw_string, "raw")                                                        \
   V(read_host_object_string, "_readHostObject")                               \
   V(readable_string, "readable")                                              \
@@ -268,6 +270,7 @@ namespace node {
   V(pipe_constructor_template, v8::FunctionTemplate)                          \
   V(process_object, v8::Object)                                               \
   V(promise_reject_function, v8::Function)                                    \
+  V(promise_wrap_template, v8::ObjectTemplate)                                \
   V(push_values_to_array_function, v8::Function)                              \
   V(randombytes_constructor_template, v8::ObjectTemplate)                     \
   V(script_context_constructor_template, v8::FunctionTemplate)                \
```

</details>

This is not optimal and could definitely at least be DRYed up. We may also want to tweak this a bit depending on what kind of access pattern we expect (people won’t usually look at `.promise` and `.parent`… right?).

JiaLiPassion commented:
@addaleax , thank you for the `resource_object` code, so if it will be implemented in your way, then in `init` hook, we can access `parent promise` with 

```javascript
  function init(id, type, triggerId, handle) {
     if (type === 'PROMISE') {
       const parentPromise = handle.parent;
         if (!parentPromise) {
            // this a new Promise 
         } else {  
            // this is a chained promise
         }
     }
  }
```

Is that correct?

AndreasMadsen commented:
@hayes Thanks for the explanation. I misunderstood what you meant by `triggerId`. Please see https://github.com/nodejs/node/pull/13000#discussion_r116296676

Plese create an issue, but don't do it around the idea the `triggerId` is wrong. Just around the idea that we don't have the `kResolve` timing information.

> quick point of clarification, setting the triggered as I am proposing here does not prevent any existing behavior, it simply adds the ability to link the execution of a callback back to the async resource that caused it to be scheduled.

I'm pretty sure it does. But we can discuss that in the issue.

AndreasMadsen commented:
# Please keep this issue on track, meaning fixing `triggerId`

@hayes will create issues for the other problems. If something is missing you are welcome to create a new issue.

JiaLiPassion commented:
@AndreasMadsen , I have updated the test cases and assert message.
https://github.com/nodejs/node/pull/13367#discussion_r119972250

please review.

addaleax commented:
@JiaLiPassion Yes, that is how you’d check whether a parent promise exists. (Now that you mention it: Maybe `handle.parent` should refer to the parent PromiseWrap, not the parent promise itself. I’m not sure.)

JiaLiPassion commented:
@addaleax , got it , thank you very much!!!

Would you please make another PR to implement it?

addaleax commented:
> Would you please make another PR to implement it?

@JiaLiPassion I’ll wait until you land this PR, then I can do that. :)

addaleax commented on test/async-hooks/test-promise.chain-promise-before-init-hooks.js:
```diff
@@ -0,0 +1,41 @@
+const initHooks = require('./init-hooks');
+const { checkInvocations } = require('./hook-checks');
+
+const p = (new Promise(common.mustCall(executor)));
```
You don’t need the outer parentheses here

JiaLiPassion replied:
yes, I will remove it.

addaleax commented on test/async-hooks/test-promise.chain-promise-before-init-hooks.js:
```diff
@@ -0,0 +1,41 @@
+function afterresolution(val) {
+  assert.strictEqual(val, 5);
+  return val;
+}
```
I think this would be a bit more readable if you just inlined `executor` and `afterresolution`

JiaLiPassion replied:
yes, you are right, I will change it.

addaleax commented on test/async-hooks/test-promise.chain-promise-before-init-hooks.js:
```diff
@@ -0,0 +1,41 @@
+}
+
+process.on('exit', onexit);
+function onexit() {
```
same for `onexit`, using `process.on('exit', () => {` works just fine

JiaLiPassion replied:
yes, I will change it.

addaleax commented on test/async-hooks/test-promise.js:
```diff
@@ -14,7 +14,7 @@ p.then(afterresolution);
 function executor(resolve, reject) {
   const as = hooks.activitiesOfTypes('PROMISE');
-  assert.strictEqual(as.length, 1, 'one activities');
+  assert.strictEqual(as.length, 1, 'one activity');
```
If you want to avoid merge conflicts with #13243, you can just drop this change

JiaLiPassion replied:
@addaleax , ok , so I just revert it.

addaleax commented on test/async-hooks/test-promise.js:
```diff
@@ -46,7 +46,8 @@ function onexit() {
   assert.strictEqual(typeof a1.uid, 'number', 'uid is a number');
-  assert.strictEqual(a1.triggerId, 1, 'parent uid 1');
+  assert.strictEqual(a1.triggerId, a0.uid,
+                     'child promise should use parent uid as triggerId');
```
Yeah, I would just not use an error message here (i.e. `assert.strictEqual(a1.triggerId, a0.uid);`). Sorry the tests are written this way.

JiaLiPassion replied:
ok, so I just modify it to 

```javascript
assert.strictEqual(a1.triggerId, a0.uid);
```

addaleax commented:
Left a few style nits.

CI: https://ci.nodejs.org/job/node-test-commit/10333/

AndreasMadsen commented:
I will fix these, there are some other changes I would like.

JiaLiPassion commented:
@AndreasMadsen , ok, thank you!  so I don't to update the source which were reviewed by @addaleax ?

AndreasMadsen commented:
@JiaLiPassion yep, I fixed it for you :)

I didn't really understand why the `triggerId` was what it was, so I checked out your code and investigated it. Nothing was wrong, but I made it a bit more robust for future changes.

JiaLiPassion commented:
@AndreasMadsen , thank you very much!!! you and @addaleax give me great help!!!

addaleax commented:
Landed in 8741e3c7501cbe66696bd3fad4a0a923e76ba11d :tada:

@JiaLiPassion I believe this is your first commit to core, if so, welcome on board! Thank you for the work you have been putting into this, and I think I speak for everyone in this thread if I say that we’d love to see more of that. If you’re interested, we can probably add you to the @nodejs/async_hooks review team that gets pinged on issues and PRs related to `async_hooks`.

@jasnell I’ve left you off as a reviewer because the PR has undergone quite some change since you reviewed, and wasn’t really ready at that point anyway. I hope that’s okay.

JiaLiPassion commented:
@addaleax , Thank you so much for all the great help, I am so glad to join here and definitely want to do more in the future.

>  If you’re interested, we can probably add you to the @nodejs/async_hooks review team that gets @mentioned on issues and PRs related to async_hooks. 

Yes!!! I am very happy to join here and it will be my honor to join @nodejs/async_hooks review team to do more contributions and want to learn more from here. 

Thank you and @AndreasMadsen again for all the great help!!!
